bundles/auth/org.eclipse.smarthome.auth.oauth2client/src/main/java/org/eclipse/smarthome/auth/oauth2client/internal/OAuthClientServiceImpl.java

```diff
@@ -0,0 +1,356 @@
+        checkNotNull(redirectURLwithParams, "redirectURLwithParams must not be null");
+
+        // parse the redirectURL
+        try {
+            URL redirectURLObject = new URL(redirectURLwithParams);
+            UrlEncoded urlEncoded = new UrlEncoded(redirectURLObject.getQuery());
+
+            String stateFromRedirectURL = urlEncoded.getValue(STATE, 0); // may contain multiple...
+            if (stateFromRedirectURL == null) {
+                if (persistedParams.state == null) {
+                    // This should not happen as the state is usually set
+                    return urlEncoded.getValue(CODE, 0);
+                } // else
+                throw new OAuthException(String.format("state from redirectURL is incorrect.  Expected: %s Found: %s",
+                        persistedParams.state, stateFromRedirectURL));
+            } else {
+                if (stateFromRedirectURL.equals(persistedParams.state)) {
+                    return urlEncoded.getValue(CODE, 0);
+                } // else
+                throw new OAuthException(String.format("state from redirectURL is incorrect.  Expected: %s Found: %s",
+                        persistedParams.state, stateFromRedirectURL));
+            }
+        } catch (MalformedURLException e) {
+            throw new OAuthException("Redirect URL is malformed", e);
+        }
+    }
+
+    @SuppressWarnings("null")
+    @Override
+    public AccessTokenResponse getAccessTokenResponseByAuthorizationCode(@NonNull String authorizationCode,
+            @NonNull String redirectURI) throws OAuthException, IOException, OAuthError {
+        OAuthConnector connector = new OAuthConnector(httpClientFactory);
+
+        if (persistedParams.redirectUri != null && !persistedParams.redirectUri.equals(redirectURI)) {
+            // check parameter redirectURI in #getAuthorizationUrl are the same as given
+            throw new OAuthException(String.format(
+                    "redirectURI should be the same from previous call #getAuthorizationUrl.  Expected: %s Found: %s",
+                    persistedParams.redirectUri, redirectURI));
+        }
+        AccessTokenResponse accessTokenResponse = connector.grantTypeAuthorizationCode(persistedParams.tokenUrl,
+                authorizationCode, persistedParams.clientId, persistedParams.clientSecret, redirectURI,
+                persistedParams.supportsBasicAuth);
+
+        if (accessTokenResponse != null) {
+            // store it
+            storeHandler.saveAccessTokenResponse(handle, accessTokenResponse);
+        }
+        return accessTokenResponse;
+    }
+
+    @SuppressWarnings("null")
+    @Override
+    public AccessTokenResponse getAccessTokenByResourceOwnerPasswordCredentials(@NonNull String username,
+            @NonNull String password, @Nullable String scope) throws OAuthException, IOException, OAuthError {
+        checkNotNull(username, "username must not be null");
```

Comment:
Since @NonNull has already been added, checkNotNull(..) precondition is not required